### DISCUSSION TITLE
What how to treat as source

### DISCUSSION THREAD

**Discussion by mwpowellhtx** [OrchardCMS/OrchardCore#11204] [Q&A]
So, more of a general CMS question than `Orchard`, per se, I expect. _What_ (necessarily, _'how'_) do we treat as _"source"_.

Obviously, any actual `CSharp` code informing the _infrastructure_, documentation, source source, resources, certificates, assemblies, etc.

But what about for the _content_ itself? We need to perform backups of the database backing the developer inner/outer and/or author inner/outer loops? Assuming that there is some sort of database versioning going on? The usual database administrator tooling fits the bill in this instance?

Will follow up in a separate question re: _author inner/outer loops_, when/where are publish operations taking place. From the _VS_ solution? Or from an _admin_ or _author_ session to the content side of things?

If either or both of these are dups, apologies. Happy to shelf if there are already responses along these lines.

Appreciate the feedback, thanks...

**Comment by Skrypt** [comment]
Content is versioned in the database with a ContentItemVersionId. So, for example, if a developer uses Liquid templates stored in the database they will be versioned to be able to revert from a previous version. Of course, right now there is no UI to do so but at least we version the Content so that it is never really also "deleted". Yes, like with any other CMS some parts of the actual content are using "short codes" or "Liquid" or scripting to code some parts of the actual website from "runtime". That means that you will want to have a database backup plan to make sure that you don't lose some of these. Of course, there is always the Import/Export feature that will allow you to also make a backup recipe of the entire website with or without actual content data.

The usual tooling for managing a database should be enough to do the database schema + data backups. For example, I use SQL Server Management Studio and Azure Data Studio to do these backups periodically. Then I add them to a repository to keep a history of these.

The publish takes place when you do a "dotnet publish" to publish the actual source code of your solution. And also you can publish content from the content side of things because that's how the Content Management in Orchard Core works. You can publish or draft content items, knowing that they can also be versioned that way. So the "publish" concept in OC is about source code and content. Though, publishing source code is something a programmer will do. And publishing content is something that an author will do.

**Reply by mwpowellhtx** [reply]
Sounds about right. I am planning on running with `postgres` for our database, analog principles there, I expect, as for `SQL Server`, etc. Good to know re: backup plans, etc. From sort of local database _dev_ authoring to perhaps _live_ authoring, though I think that is probably avoidable if we can contain authoring to a local _dev_ or _staging_ instance. Thanks for the feedback @Skrypt. Anyone else with any other insights, thanks in advance.

**Reply by Skrypt** [reply]
Yes, If more than one person is working on the website the best approach is to use a staging environment for the database. Docker can sometimes be useful for this. Try to avoid using the prod database directly and deploy the data by using the import/export feature if you can. Some use it and find it is good enough for their need. I personally prefer using pure SQL for doing data migrations from one database to another. SQL Data compare, SQL Schema compare.

**Comment by mwpowellhtx** [comment]
This I can envision being framed in terms of workflow, pushing and pulling various records between databases, _dev_ to/from _staging_, pushing from _dev_ or _staging_ to _prod_, and so on. There is adequate separation of the various _content item_, _composite_ records that this should be possible?
